### PR TITLE
Refactor tunnel

### PR DESCRIPTION
**Pull request by NielsZeilemaker**
Supercedes #1117, rebased on devel fixes hidden tunnels

### CODE REVIEW

**Comment by NielsZeilemaker** [comment]
@whirm the new tunnel crypto depens on cryptography. It's a package that exposes the gcm mode of openssl.

File: README.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ We make use of submodules, so remember using the --recursive argument when cloni
 
 ### Debian/Ubuntu/Mint
 ```bash
-sudo apt-get install libav-tools libjs-excanvas libjs-mootools libx11-6 python-apsw python-cherrypy3 python-crypto python-feedparser python-gmpy python-libtorrent python-m2crypto python-netifaces python-pil python-pyasn1 python-requests python-twisted python-wxgtk2.8 python2.7 vlc
+sudo apt-get install libav-tools libjs-excanvas libjs-mootools libx11-6 python-apsw python-cherrypy3 python-feedparser python-gmpy python-libtorrent python-m2crypto python-netifaces python-pil python-requests python-twisted python-wxgtk2.8 python2.7 vlc
```
Comment (whirm) [inline README.md]: @NielsZeilemaker could you add python-cryptography? also in ```debian/control```

Comment (LipuFei) [inline README.md]: I guess it's `python-crypto`.

Comment (NielsZeilemaker) [inline README.md]: @LipuFei, its python-cryptography python-crypto was used before. I'll add the package

**Comment by whirm** [comment]
@NielsZeilemaker I've updated the gorillas, let's see how it goes.

**Comment by whirm** [comment]
retest this please

**Comment by NielsZeilemaker** [comment]
retest this please

**Comment by whirm** [comment]
seems that it got executed while I was installing the package :)

**Comment by synctext** [comment]
Almost :-)
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/usr/lib/python2.7/dist-packages/nose/loader.py", line 411, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/Test/test_tunnel_community.py", line 16, in <module>
    from Tribler.community.tunnel.tunnel_community import TunnelSettings
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/community/tunnel/tunnel_community.py", line 27, in <module>
    from Tribler.community.tunnel.crypto.tunnelcrypto import TunnelCrypto, CryptoException
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/community/tunnel/crypto/tunnelcrypto.py", line 3, in <module>
    from cryptowrapper import crypto_box_beforenm, crypto_auth, crypto_auth_verify, Cipher, algorithms, modes, HKDFExpand, hashes, default_backend
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/community/tunnel/crypto/cryptowrapper.py", line 14, in <module>
    from cryptography.hazmat.backends import default_backend
ImportError: No module named cryptography.hazmat.backends
```

**Comment by synctext** [comment]
Ai. Assert:
http://jenkins.tribler.org/job/GH_Tribler_pull-request-tester_devel/3324/artifact/output//Screenshot-35.png
Multiple Tracebacks
```
DEBUG   1422873500.43        dispersy:778   ignore vote for ('95.211.198.142', 49165) from ('95.211.198.142', 44738) (voter is within our LAN)
DEBUG   1422873500.43       community:2640  received introduction to {95.211.198.142:22872} from {95.211.198.142:44738}
ERROR   1422873500.43       community:2184  exception during check_callback for dispersy-puncture
Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/dispersy/community.py", line 2182, in on_messages
    possibly_messages = list(meta.check_callback(messages))
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/dispersy/community.py", line 2865, in check_puncture
    identifiers_seen[message.payload.identifer] = message.candidate
AttributeError: 'Implementation' object has no attribute 'identifer'
DEBUG   1422873500.43       community:2011  got 1 incoming packets
```
```
INFO    1422873504.39 tunnel_community:420   could not create circuit, no available hops.
DEBUG   1422873504.39   SocketHandler:503   SocketHandler: Got UDP data ('95.211.198.142', 49165) len 166
DEBUG   1422873504.39        endpoint:72                              cell <-  95.211.198.142:49165  166 bytes
DEBUG   1422873504.39       community:2011  got 1 incoming packets
DEBUG   1422873504.39       community:2213  in... 1 cell messages from {95.211.198.142:49165}
DEBUG   1422873504.39 tunnel_community:751   Got create (4076263781) from ('95.211.198.142', 49165)
ERROR   1422873504.39        dispersy:1723  exception during handle_callback for cell
Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/dispersy/dispersy.py", line 1718, in _update
    messages[0].handle_callback(messages)
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/community/tunnel/tunnel_community.py", line 766, in on_cell
    self.dispersy.on_incoming_packets([(message.candidate, TunnelConversion.convert_from_cell(packet))], False, source=u"circuit_%d" % circuit_id)
TypeError: on_incoming_packets() got an unexpected keyword argument 'source'
DEBUG   1422873504.39       community:2011  got 1 incoming packets
DEBUG   1422873504.39       community:2213  in... 1 cell messages from {95.211.198.142:49165}
DEBUG   1422873504.39 tunnel_community:751   Got create (3294018314) from ('95.211.198.142', 49165)
```

**Comment by NielsZeilemaker** [comment]
I guess it needs a new dispersy pointer, let me have a look to see if I can push one in.

**Comment by NielsZeilemaker** [comment]
I'll have to do it tonight as I don't have access to my laptop, alternatively @whirm can you send me a pull request with the new dispersy pointer?

**Comment by synctext** [comment]
@whirm please help out Niels with that pull request.
If it passes then master student @rjruigrok can start testing matters full-time.

**Comment by NielsZeilemaker** [comment]
The tunnel part passed, but two of the download tests didn't I'll retest it

**Comment by NielsZeilemaker** [comment]
retest this please

**Comment by synctext** [comment]
One TraceBack is going: unexpected keyword argument 'source'
One left:
```
DEBUG   1422878751.43       community:2011  got 1 incoming packets
ERROR   1422878751.43       community:2184  exception during check_callback for dispersy-puncture
Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/dispersy/community.py", line 2182, in on_messages
    possibly_messages = list(meta.check_callback(messages))
  File "/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/dispersy/community.py", line 2865, in check_puncture
    identifiers_seen[message.payload.identifer] = message.candidate
AttributeError: 'Implementation' object has no attribute 'identifer'
DEBUG   1422878751.43       community:2011  got 1 incoming packets
```

**Comment by NielsZeilemaker** [comment]
It's a small bug in dispersy devel, identifer should be identifier

**Comment by whirm** [comment]
I'll fix it

**Comment by whirm** [comment]
yay, @NielsZeilemaker is it ready to merge?

**Comment by NielsZeilemaker** [comment]
Let me have a quick look tonight at the hidden_tunnel stuffs and then you can. I remember removing some code which closed the exit-socket of the tunnels which get linked, which I should put back.
Moreover, let's have @rjruigrok have a look before we merge it. Or do you want to merge it first and have a look later?

**Comment by whirm** [comment]
@NielsZeilemaker well, it's always better to check before than after, but as the stuff is broken on devel anyway. I don't really care.

Also, don't forget to update ```debian/control```

**Comment by NielsZeilemaker** [comment]
I'll push that change tonight as well.

**Comment by synctext** [comment]
yay2, @NielsZeilemaker!  Easier if merged into devel and then @rjruigrok can check it further.

**Comment by rjruigrok** [comment]
Nice work! Refactored parts look good.

Just managed to get it working, dependencies libsodium and python-cryptography where not yet available in Ubuntu 14.10, but built them from source and works - partly. I suspect I'm using the wrong python-cryptography here.  

  File "/home/rob/git/tribler/Tribler/community/tunnel/tunnel_community.py", line 681, in _ours_on_created_extended
    shared_secret = self.crypto.verify_and_generate_shared_secret(hop.dh_secret, message.payload.key, message.payload.auth, hop.public_key.key.pk)
  File "/home/rob/git/tribler/Tribler/community/tunnel/crypto/tunnelcrypto.py", line 39, in verify_and_generate_shared_secret
    crypto_auth_verify(auth, dh_received, shared_secret)
TypeError: crypto_auth_verify() takes exactly 2 arguments (3 given)

**Comment by NielsZeilemaker** [comment]
That's the old version of libnacl, update your submodule recursively.

**Comment by NielsZeilemaker** [comment]
@whirm it's ready

@rjruigrok have a look at the community, I would pay special attention to the announce stuff. I'm not sure what will happen if a introduction_point of a seeder closes. Will the seeder open a new tunnel?, are you sure this will always happen? Maybe add a periodic task which creates a new introduction_point after X seconds (similar to the dht_lookup which is scheduled every 60s).

**Pull request merged**